Model/Service/Sync/AsyncBulkConsumer.php

```diff
@@ -0,0 +1,140 @@
+ *
+ * Class AsyncBulkConsumer
+ */
+class AsyncBulkConsumer
+{
+    /** @var Logger */
+    private $logger;
+
+    /** @var JsonHelper */
+    private $jsonHelper;
+
+    /** @var IndexRepository */
+    private $indexRepository;
+
+    /** @var NostoSyncService */
+    private $nostoSyncService;
+
+    /** @var NostoScopeHelper */
+    private $nostoScopeHelper;
+
+    /** @var EntityManager */
+    private $entityManager;
+
+    /**
+     * AsyncBulkConsumer constructor.
+     *
+     * @param Logger $logger
+     * @param JsonHelper $jsonHelper
+     * @param IndexRepository $indexRepository
+     * @param NostoSyncService $nostoSyncService
+     * @param NostoScopeHelper $nostoScopeHelper
+     * @param EntityManager $entityManager
+     */
+    public function __construct(
+        Logger $logger,
+        JsonHelper $jsonHelper,
+        IndexRepository $indexRepository,
+        NostoSyncService $nostoSyncService,
+        NostoScopeHelper $nostoScopeHelper,
+        EntityManager $entityManager
+    ) {
+        $this->logger = $logger;
+        $this->jsonHelper = $jsonHelper;
+        $this->indexRepository = $indexRepository;
+        $this->nostoSyncService = $nostoSyncService;
+        $this->nostoScopeHelper = $nostoScopeHelper;
+        $this->entityManager = $entityManager;
+    }
+
+    /**
+     * Processing operation for product sync
+     *
+     * @param array|\Magento\AsynchronousOperations\Api\Data\OperationInterface $operation
+     * @return void
+     * @throws Exception
+     * @suppress PhanUndeclaredClassConstant
+     */
+    public function processOperation($operation)
+    {
+        $errorCode = null;
+        $message = null;
+        if (!is_array($operation)) {
+            $serializedData = $operation->getSerializedData();
+        } else {
+            $serializedData = $operation['data']['serialized_data'];
+        }
+        $unserializedData = $this->jsonHelper->jsonDecode($serializedData);
+        $productIds = $unserializedData['product_ids'];
+        $storeId = $unserializedData['store_id'];
+        try {
+            $store = $this->nostoScopeHelper->getStore($storeId);
+            $outOfSyncCollection = $this->indexRepository->getByProductIdsAndStoreId($productIds, $storeId);
+            $this->nostoSyncService->syncIndexedProducts($outOfSyncCollection, $store);
+            $this->nostoSyncService->syncDeletedProducts($store);
+        } catch (Exception $e) {
+            $this->logger->critical($e->getMessage());
+            $message = __('Something went wrong when syncing products to Nosto. Check log for details.');
+            if (!is_array($operation)) {
+                $operation->setStatus(
+                    \Magento\AsynchronousOperations\Api\Data\OperationInterface::STATUS_TYPE_NOT_RETRIABLY_FAILED
+                )->setErrorCode($e->getCode())
+                ->setResultMessage($message);
+            }
+        }
+        if (!is_array($operation)) {
```

Comment:
This should be inside `try` block. Otherwise we'll always set the operation as complete regardless if it fails or not.